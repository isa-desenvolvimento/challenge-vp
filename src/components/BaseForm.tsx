import { Formik, Form } from 'formik'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const BaseForm = ({ initialValues, baseForm, onSubmit }: any) => {
  const renderForm = () => {
    return (
      <Formik
        onSubmit={(data, action) => onSubmit(data, action)}
        initialValues={initialValues}
      >
        {() => <Form>{baseForm}</Form>}
      </Formik>
    )
  }

  return <>{renderForm()}</>
}
