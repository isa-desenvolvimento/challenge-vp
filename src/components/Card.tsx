import {
  StyleCardContainer,
  StyleCardContent,
  StyleCardFooter,
  StyleCardHeader,
} from '@/styles/Card'
import { CardProps } from '@/types/global'
import React from 'react'

export const Card = ({
  header,
  content,
  footer,
  primary,
  display,
  height,
}: CardProps) => {
  const renderHeader = () =>
    header && <StyleCardHeader primary={primary}>{header}</StyleCardHeader>
  const renderContent = () =>
    content && (
      <StyleCardContent primary={primary} height={height}>
        {content}
      </StyleCardContent>
    )
  const renderFooter = () =>
    footer && <StyleCardFooter primary={primary}>{footer}</StyleCardFooter>

  return (
    <StyleCardContainer display={display}>
      {renderHeader()}
      {renderContent()}
      {renderFooter()}
    </StyleCardContainer>
  )
}
