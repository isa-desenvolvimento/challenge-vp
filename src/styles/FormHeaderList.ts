import styled from 'styled-components'

export const StyleHeadList = styled.div`
  display: grid;
  grid-auto-columns: 40%;
  padding: 0.5rem;

  @media screen and (max-width: 600px) {
    display: revert;
    width: 100%;
  }
`
