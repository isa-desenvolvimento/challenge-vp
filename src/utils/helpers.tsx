import { render } from '@testing-library/react'
import GlobalContext from '../context'

export const renderWithProvider = (children: React.ReactNode) => {
  render(<GlobalContext>{children}</GlobalContext>)
}
