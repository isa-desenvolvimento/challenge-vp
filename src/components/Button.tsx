import { Icon } from '@/components/Icon'
import { StyleButton } from '@/styles/Button'
import { ButtonProps } from '@/types/global'

export const Button = ({
  iconProps,
  primary,
  onClick,
  children,
  size,
  type,
}: ButtonProps) => {
  const renderIcon = () => iconProps && <Icon faIcon={iconProps}></Icon>

  return (
    <StyleButton
      role="button"
      arial-label="button"
      primary={primary}
      size={size}
      type={type}
      onClick={onClick}
    >
      {renderIcon()}
      {children}
    </StyleButton>
  )
}
