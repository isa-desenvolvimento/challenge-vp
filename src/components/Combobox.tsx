import { ComboboxProps } from '@/types/global'

import { DropdownWrapper, StyledSelect, StyledOption } from '@/styles/Combobox'
import { useField } from 'formik'

export const Combobox = ({
  role,
  options,
  placeholder,
  primary,
  onChange,
  name,
  isFormik,
}: ComboboxProps) => {
  const [field, meta] = useField(name)

  const renderSelection = () => {
    if (isFormik) {
      return (
        <StyledSelect primary={primary} {...field} {...meta} role={role}>
          <StyledOption selected value="" name={`${name}.defaultValue`}>
            {placeholder}
          </StyledOption>
          {renderOptions()}
        </StyledSelect>
      )
    }

    return (
      <StyledSelect onChange={onChange} primary={primary}>
        <StyledOption selected value="" name={`${name}.defaultValue`}>
          {placeholder}
        </StyledOption>
        {renderOptions()}
      </StyledSelect>
    )
  }

  const renderOptions = () => {
    return (
      options?.length &&
      options.map(({ value }) => {
        return (
          <StyledOption name={`${name}.${value}`} key={value} value={value}>
            {value}
          </StyledOption>
        )
      })
    )
  }

  return <DropdownWrapper role="combobox">{renderSelection()}</DropdownWrapper>
}
