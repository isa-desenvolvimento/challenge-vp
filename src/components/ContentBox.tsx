import { StyleContainer } from '@/styles/ContentBox'
import { ChildrenProps } from '@/types/global'

export const ContentBox = ({ children }: ChildrenProps) => {
  return <StyleContainer role="contentBox"> {children}</StyleContainer>
}
