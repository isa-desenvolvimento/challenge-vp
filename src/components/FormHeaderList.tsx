import { Combobox } from '@/components/Combobox'
import { BaseForm } from './BaseForm'
import { StyleHeadList } from '@/styles/FormHeaderList'
import { SearchContainerProps } from '@/types/global'

export const FormHeaderList = ({ options, onSubmit }: SearchContainerProps) => {
  const renderBaseForm = () => (
    <StyleHeadList>
      <Combobox
        options={options}
        placeholder="Status"
        name="orderby"
        onChange={onSubmit}
      />
    </StyleHeadList>
  )

  return (
    <BaseForm
      baseForm={renderBaseForm()}
      initialValues={{
        orderby: '',
      }}
    />
  )
}
