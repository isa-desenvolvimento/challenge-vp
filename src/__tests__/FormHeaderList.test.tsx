import { renderWithProvider } from '../utils/helpers'
import { screen } from '@testing-library/react'

import { FormHeaderList } from '@/components/FormHeaderList'

describe('<FormHeaderList />', () => {
  const optionsHeaderList = [
    { value: 'Title', id: 1 },
    { value: 'Description', id: 2 },
    { value: 'Status', id: 3 },
  ]
  const onSubmit = jest.fn()

  beforeEach(() =>
    renderWithProvider(
      <FormHeaderList options={optionsHeaderList} onSubmit={onSubmit} />
    )
  )

  it('should render FormHeaderList', () => {
    expect(screen.getByRole(/form-header-list/i).textContent).toBeTruthy()
  })

  it('should render the styled-components tag-status', () => {
    expect(screen.getByRole(/form-header-list/i)).toHaveStyle({
      display: 'grid',
      'grid-auto-columns': '40%',
      padding: '0.5rem',
    })
  })

  it('should render snapshot', () => {
    expect(screen.getByRole(/form-header-list/i)).toMatchSnapshot()
  })
})
