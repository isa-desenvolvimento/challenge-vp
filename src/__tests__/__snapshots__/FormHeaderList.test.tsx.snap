// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormHeaderList /> should render snapshot 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  gap: 1rem;
  padding: 1rem;
  text-overflow: ellipsis;
  background: #ffff;
  color: #3AA3AA;
  border: none;
  border-radius: 5px;
  width: auto;
}

.c3 {
  background: #eceff5;
  color: #3AA3AA;
}

.c0 {
  display: grid;
  grid-auto-columns: 40%;
  padding: 0.5rem;
}

@media screen and (max-width:600px) {
  .c2 {
    margin: 1rem;
  }
}

@media screen and (max-width:600px) {
  .c0 {
    display: revert;
    width: 100%;
  }
}

<div
  class="c0"
  role="form-header-list"
>
  <div
    class="c1"
    role="combobox"
  >
    <select
      class="c2"
    >
      <option
        class="c3"
        name="orderby.defaultValue"
        value=" "
      >
        Status
      </option>
      <option
        class="c3"
        name="orderby.Title"
        value="Title"
      >
        Title
      </option>
      <option
        class="c3"
        name="orderby.Description"
        value="Description"
      >
        Description
      </option>
      <option
        class="c3"
        name="orderby.Status"
        value="Status"
      >
        Status
      </option>
    </select>
  </div>
</div>
`;
